### ISSUE TITLE
Getting libuv event loop fails in Electron main process on Windows

### ISSUE THREAD

**Issue by laptou** [node-ffi-napi/get-uv-event-loop-napi-h#3]
I opened an issue in the main repository here: https://github.com/node-ffi-napi/node-ffi-napi/issues/20

But the offending code actually seems to be in this repository, right here: https://github.com/node-ffi-napi/get-uv-event-loop-napi-h/blob/cb87f0536d26401244da5064fb529f315ac48e01/include/get-uv-event-loop-napi.h#L17

Whenever I attempt to use `ffi-napi` in an Electron main process and call an async function, this assertion fails and crashes the whole application.

**Comment by addaleax** [comment]
Could you check whether this solves your issue?

```diff
diff --git a/include/get-uv-event-loop-napi.h b/include/get-uv-event-loop-napi.h
index 9ee1f472e9f4..2f2797d888bd 100644
--- a/include/get-uv-event-loop-napi.h
+++ b/include/get-uv-event-loop-napi.h
@@ -1,6 +1,11 @@
 #ifndef _GET_UV_EVENT_LOOP_NAPI_H
+#define _GET_UV_EVENT_LOOP_NAPI_H
 
+#if !defined(NAPI_VERSION) || NAPI_VERSION < 2
 #include <get-symbol-from-current-process.h>
+#endif
+
+#include <node_api.h>
 
 inline
 uv_loop_t* get_uv_event_loop(napi_env env) {
@@ -12,8 +17,12 @@ uv_loop_t* get_uv_event_loop(napi_env env) {
   assert(status == napi_ok);
   if (napi_version < 2)
     return uv_default_loop();
+#if !defined(NAPI_VERSION) || NAPI_VERSION < 2
   napi_get_uv_event_loop__ = (get_uv_event_loop_fn)
       get_symbol_from_current_process("napi_get_uv_event_loop");
+#else
+  napi_get_uv_event_loop__ = &napi_get_uv_event_loop;
+#endif
   assert(napi_get_uv_event_loop__ != NULL);
   uv_loop_t* ret;
   status = napi_get_uv_event_loop__(env, &ret);
```

**Comment by laptou** [comment]
Revised the file to look like [this](https://gist.github.com/laptou/b518ca6c3e541b873942ab2c7b8d1e82), did not work. I rebuilt it using `npx electron-rebuild -f`, and the assertion is now on line 26, which indicates that the changes went through, but did not solve the problem 😕

![image](https://user-images.githubusercontent.com/14832331/47607957-f1655e00-d9f4-11e8-82fe-839c5cb7b0e6.png)

**Comment by addaleax** [comment]
Hm … that’s odd – can you check which value `NAPI_VERSION` has (after including `node_api.h`)?

**Comment by laptou** [comment]
`napi_version` is 3, `NAPI_VERSION` appears to be undefined. I simply added this code after line 14:

```
#ifdef NAPI_VERSION
  printf("%u", NAPI_VERSION);
#endif
```

And placed a breakpoint inside of it. The breakpoint is never hit, and nothing appears on stdout. I inspected `napi_version` by attaching to the process with Visual Studio.

After looking at this for a couple of minutes, I think I understand what is happening -- the new code isn't being executed, because `NAPI_VERSION` is undefined. I looked at the header files at `~/.electron-gyp/.node-gyp/iojs-3.0.6/src/node_api.h` and I think I've found the answer:

Lines 10-18 of [this file](https://github.com/nodejs/node/blob/336651bad0368e1a4ea5597c9a0615764dba4d16/src/node_api.h#L10) are missing from that file.

Once I added them, then `electron-rebuild` began to produce errors like these:

> `c:\users\...\ffi-napi-test\node_modules\get-uv-event-loop-napi-h\include\get-uv-event-loop-napi.h(22): error C3861: 'assert': could not find the identifier (compiling file from ..\src\ffi.cc) [C:\Users\...\ffi-napi-test\node_modules\ffi-napi\build\ffi_bindings.vcxproj]`

assert.h was being included by get-symbol-from-current-process.h, and since NAPI_VERSION was now defined, get-symbol-from-current-process.h was no longer being included.

I added an `#include <assert.h>` myself to the top of `get-uv-event-loop-napi.h`, and voila -- the problem is gone.

## Summary

Here's a [gist](https://gist.github.com/laptou/3bb7fe4fe49a9a63d72f41820924efbe) with all of the relevant files. In short, the io.js version of the N-API header doesn't contain the `NAPI_VERSION` macro, [because this was added in Node.js v11](https://github.com/nodejs/node/commit/8476053c132fd9613aab547aba165190f8064254#diff-50f0ab3fe6127f546fbe8aabab5a0d7c) and Electron uses Node.js v10 causing `get-uv-event-loop-napi.h` to attempt to use `get_symbol_from_current_process` to obtain reference to the function -- which did not work because our N-API version is 3.

**Comment by laptou** [comment]
released in 1.0.4, thanks
